crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -1230,6 +1234,135 @@ impl ParallelSparseTrie {
         nodes.extend(self.upper_subtrie.nodes.iter());
         nodes
     }
+
+    /// Example implementation of parallel node revealing
+    /// TODO move this onto the trait
+    pub fn reveal_nodes(&mut self, mut nodes: Vec<RevealedSparseNode>) -> SparseTrieResult<()> {
+        if nodes.is_empty() {
+            return Ok(())
+        }
+
+        // Sort nodes first by their subtrie, and secondarily by their path. This allows for
+        // grouping nodes by their subtrie using `chunk_by`.
+        nodes.sort_unstable_by(
+            |RevealedSparseNode { path: path_a, .. }, RevealedSparseNode { path: path_b, .. }| {
+                let subtrie_type_a = SparseSubtrieType::from_path(path_a);
+                let subtrie_type_b = SparseSubtrieType::from_path(path_b);
+                subtrie_type_a.cmp(&subtrie_type_b).then(path_a.cmp(path_b))
+            },
+        );
+
+        // Update the top-level branch node masks. This is simple and can't be done in parallel.
+        for RevealedSparseNode { path, masks, .. } in nodes.iter() {
+            if let Some(tree_mask) = masks.tree_mask {
+                self.branch_node_tree_masks.insert(*path, tree_mask);
+            }
+            if let Some(hash_mask) = masks.hash_mask {
+                self.branch_node_hash_masks.insert(*path, hash_mask);
+            }
+        }
+
+        let mut nodes_slice = nodes.as_slice();
+
+        // Due to the sorting, if there are any nodes to be revealed in the upper trie then the
+        // first node will be an upper trie node. Handle all upper trie reveals using
+        // `ParallelSparseTrie::reveal_node`, which deals with revealing recursively into
+        // lower-subtries as necessary.
+        if SparseSubtrieType::from_path(&nodes[0].path).lower_index().is_none() {
+            let mut num_upper_nodes = 0;
+            let upper_nodes =
+                nodes.iter().take_while(|n| SparseSubtrieType::path_len_is_upper(n.path.len()));
+
+            for node in upper_nodes {
+                // TODO once reveal_node is not on the trait we can probably make it accept
+                // references, and skip these clones.
+                self.reveal_node(node.path, node.node.clone(), node.masks.clone())?;
+                num_upper_nodes += 1;
+            }
```

Comment:
I think we can do something like this
```suggestion
    // find where upper nodes end
    let split_pos = nodes
        .iter()
        .position(|n| !SparseSubtrieType::path_len_is_upper(n.path.len()))
        .unwrap_or(nodes.len());
    
    // get owned nodes by draining
    for node in nodes.drain(..split_pos) {
        self.reveal_node(node.path, node.node, node.masks)?;
    }
```
and we can create the slice after

Reply:
I think draining off the front of the Vec will cause the rest of the elements to get shifted left, so that's incurring a potentially large copy if there's a lot of proofs. The original way also only iterates over the upper nodes once, not twice, so I think overall will be more performant (though admittedly much uglier)

Reply:
I think my approach with iterator also iterates only once and doesn't copy? Or does it not work for some reason?

Reply:
Sorry @shekhirin , GH hadn't updated with your comment and I didn't see it 🤦

I ended up finding yet a third approach, which is kind of a mix of me and Dan's. It allows for reserving capacity on the upper trie's nodes map without introducing any new allocs. We're back to two iterations over upper nodes though... can't have everything 🤷